File: android/src/main/java/org/tiqr/authenticator/MainActivity.java
package org.tiqr.authenticator;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.view.View;
import android.webkit.WebView;

import com.google.android.c2dm.C2DMessaging;

import org.tiqr.authenticator.auth.AuthenticationChallenge;
import org.tiqr.authenticator.auth.EnrollmentChallenge;
import org.tiqr.authenticator.authentication.AuthenticationActivityGroup;
import org.tiqr.authenticator.datamodel.DbAdapter;
import org.tiqr.authenticator.dialog.ActivityDialog;
import org.tiqr.authenticator.enrollment.EnrollmentActivityGroup;
import org.tiqr.authenticator.general.HeaderView;
import org.tiqr.authenticator.qr.CaptureActivity;
import org.tiqr.service.authentication.AuthenticationService;
import org.tiqr.service.authentication.ParseAuthenticationChallengeError;
import org.tiqr.service.enrollment.EnrollmentService;
import org.tiqr.service.enrollment.ParseEnrollmentChallengeError;
import org.tiqr.service.notification.NotificationService;

import javax.inject.Inject;

public class MainActivity extends Activity {

    public static final int MY_PERMISSIONS_REQUEST_CAMERA = 42;


    @Inject
    NotificationService _notificationService;

    protected
    @Inject
    EnrollmentService _enrollmentService;

    protected
    @Inject
    AuthenticationService _authenticationService;

    protected
    @Inject
    DbAdapter _dbAdapter;

    private ActivityDialog activityDialog;


    /**
     * Called when the activity is first created.
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);
        ((Application)getApplication()).inject(this);

        HeaderView headerView = (HeaderView)findViewById(R.id.headerView);
        headerView.hideLeftButton();

        int contentResource = 0;
        if (_dbAdapter.identityCount() > 0) {
            contentResource = R.string.main_text_instructions_for_pincode;
        } else {
            headerView.hideRightButton();
            contentResource = R.string.main_text_welcome;
        }

        loadContentsIntoWebView(contentResource, R.id.webview);

        findViewById(R.id.scan_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (ContextCompat.checkSelfPermission(MainActivity.this, android.Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
                    _startScanning();
                } else {
                    ActivityCompat.requestPermissions(MainActivity.this, new String[]{android.Manifest.permission.CAMERA}, MY_PERMISSIONS_REQUEST_CAMERA);
                }
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
        switch (requestCode) {
            case MY_PERMISSIONS_REQUEST_CAMERA: {
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    _startScanning();
                }
            }
        }
    }

    private void _startScanning() {
        Intent scanIntent = new Intent(MainActivity.this, CaptureActivity.class);
        startActivity(scanIntent);
    }

    public void loadContentsIntoWebView(int contentResourceId, int webviewResourceId) {
        WebView webView = (WebView)findViewById(webviewResourceId);
        String data = getString(contentResourceId);
        //needed to render chars correctly
        webView.loadDataWithBaseURL(null, data, "text/html", "utf-8", null);
        webView.setBackgroundColor(Color.TRANSPARENT);
    }

    public void onStart() {
        super.onStart();

        String deviceToken = C2DMessaging.getRegistrationId(this);
        if (deviceToken != null && !"" .equals(deviceToken)) {
            _notificationService.sendRequestWithDeviceToken(deviceToken);
        } else {
            C2DMessaging.register(this, C2DMReceiver.SENDER_ID);
        }
        // Handle tiqrauth:// and tiqrenroll:// URLs
        final Intent intent = getIntent();
        final String action = intent.getAction();
        if (Intent.ACTION_VIEW.equals(action)) {
            // Perform the actions needed to parse the raw challenge
            activityDialog = ActivityDialog.show(this);
            String rawChallenge = intent.getDataString();
            if (_enrollmentService.isEnrollmentChallenge(rawChallenge)) {
                _enroll(rawChallenge);
            } else {
                _authenticate(rawChallenge);
            }
        }
    }

    /**
     * Parse authentication challenge and start authentication process.
     *
     * @param challenge The raw authentication challenge.
     */
    private void _authenticate(String challenge) {
        _authenticationService.parseAuthenticationChallenge(challenge, new AuthenticationService.OnParseAuthenticationChallengeListener() {
            @Override
            public void onParseAuthenticationChallengeSuccess(AuthenticationChallenge challenge) {
                activityDialog.cancel();
                Intent intent = new Intent(getApplicationContext(), AuthenticationActivityGroup.class);
                intent.putExtra("org.tiqr.challenge", challenge);
                intent.putExtra("org.tiqr.protocolVersion", "2");
                startActivity(intent);
            }

            @Override
            public void onParseAuthenticationChallengeError(ParseAuthenticationChallengeError error) {
                activityDialog.cancel();
                new AlertDialog.Builder(MainActivity.this)
                        .setTitle(error.getTitle())
                        .setMessage(error.getMessage())
                        .setCancelable(false)
                        .setPositiveButton(R.string.ok_button, null)
                        .show();
            }
        });
    }

    /**
     * Parse enrollment challenge and start enrollment process.
     *
     * @param challenge The raw challenge to enroll for.
     */
    private void _enroll(String challenge) {
        _enrollmentService.parseEnrollmentChallenge(challenge, new EnrollmentService.OnParseEnrollmentChallengeListener() {
            @Override
            public void onParseEnrollmentChallengeSuccess(EnrollmentChallenge challenge) {
                activityDialog.cancel();
                Intent intent = new Intent(getApplicationContext(), EnrollmentActivityGroup.class);
                intent.putExtra("org.tiqr.challenge", challenge);
                intent.putExtra("org.tiqr.protocolVersion", "2");
                startActivity(intent);
            }

            @Override
            public void onParseEnrollmentChallengeError(ParseEnrollmentChallengeError error) {
                activityDialog.cancel();

                new AlertDialog.Builder(MainActivity.this)
                        .setTitle(error.getTitle())
                        .setMessage(error.getMessage())
                        .setCancelable(false)
                        .setPositiveButton(R.string.ok_button, null)
                        .show();
            }
        });
    }
}

File: android/src/main/java/org/tiqr/authenticator/MainActivity.kt
package org.tiqr.authenticator

import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.webkit.WebView
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import org.tiqr.authenticator.auth.AuthenticationChallenge
import org.tiqr.authenticator.auth.EnrollmentChallenge
import org.tiqr.authenticator.authentication.AuthenticationActivityGroup
import org.tiqr.authenticator.base.BaseActivity
import org.tiqr.authenticator.databinding.ActivityMainBinding
import org.tiqr.authenticator.datamodel.DbAdapter
import org.tiqr.authenticator.dialog.ActivityDialog
import org.tiqr.authenticator.enrollment.EnrollmentActivityGroup
import org.tiqr.service.authentication.AuthenticationService
import org.tiqr.service.authentication.ParseAuthenticationChallengeError
import org.tiqr.service.enrollment.EnrollmentService
import org.tiqr.service.enrollment.ParseEnrollmentChallengeError
import javax.inject.Inject

class MainActivity : BaseActivity<ActivityMainBinding>() {

    @Inject
    internal lateinit var enrollmentService: EnrollmentService

    @Inject
    internal lateinit var authenticationService: AuthenticationService

    @Inject
    internal lateinit var dbAdapter: DbAdapter

    private var activityDialog: ActivityDialog? = null


    override val layout = R.layout.activity_main

    /**
     * Called when the activity is first created.
     */
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        TiqrApplication.component(this).inject(this)

        binding.header.hideLeftButton()

        val contentResource: Int
        if (dbAdapter.identityCount() > 0) {
            contentResource = R.string.main_text_instructions
        } else {
            binding.header.hideRightButton()
            contentResource = R.string.main_text_welcome
        }

        loadContentsIntoWebView(contentResource, binding.webview)

        binding.btnScan.setOnClickListener {
            if (ContextCompat.checkSelfPermission(this@MainActivity, android.Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
                startScanning()
            } else {
                ActivityCompat.requestPermissions(this@MainActivity, arrayOf(android.Manifest.permission.CAMERA), MY_PERMISSIONS_REQUEST_CAMERA)
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            MY_PERMISSIONS_REQUEST_CAMERA -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    startScanning()
                }
            }
        }
    }

    private fun startScanning() {
        // TODO: open scan activity
    }

    private fun loadContentsIntoWebView(contentResourceId: Int, webView: WebView) {
        val data = getString(contentResourceId)
        //needed to render chars correctly
        webView.loadDataWithBaseURL(null, data, "text/html", "utf-8", null)
        webView.setBackgroundColor(Color.TRANSPARENT)
    }

    public override fun onStart() {
        super.onStart()
        // Handle tiqrauth:// and tiqrenroll:// URLs
        val intent = intent
        val action = intent.action
        if (Intent.ACTION_VIEW == action) {
            // Perform the actions needed to parse the raw challenge
            activityDialog = ActivityDialog.show(this)
            val rawChallenge = intent.dataString
            if (enrollmentService.isEnrollmentChallenge(rawChallenge!!)) {
                enroll(rawChallenge)
            } else {
                authenticate(rawChallenge)
            }
        }
    }

    /**
     * Parse authentication challenge and start authentication process.
     *
     * @param challenge The raw authentication challenge.
     */
    private fun authenticate(challenge: String) {
        authenticationService.parseAuthenticationChallenge(challenge, object : AuthenticationService.OnParseAuthenticationChallengeListener {
            override fun onParseAuthenticationChallengeSuccess(challenge: AuthenticationChallenge) {
                activityDialog?.cancel()
                val intent = Intent(applicationContext, AuthenticationActivityGroup::class.java)
                intent.putExtra("org.tiqr.challenge", challenge)
                intent.putExtra("org.tiqr.protocolVersion", "2")
                startActivity(intent)
            }

            override fun onParseAuthenticationChallengeError(error: ParseAuthenticationChallengeError) {
                activityDialog?.cancel()
                AlertDialog.Builder(this@MainActivity)
                        .setTitle(error.title)
                        .setMessage(error.message)
                        .setCancelable(false)
                        .setPositiveButton(R.string.ok_button, null)
                        .show()
            }
        })
    }

    /**
     * Parse enrollment challenge and start enrollment process.
     *
     * @param challenge The raw challenge to enroll for.
     */
    private fun enroll(challenge: String) {
        enrollmentService.parseEnrollmentChallenge(challenge, object : EnrollmentService.OnParseEnrollmentChallengeListener {
            override fun onParseEnrollmentChallengeSuccess(challenge: EnrollmentChallenge) {
                activityDialog?.cancel()
                val intent = Intent(applicationContext, EnrollmentActivityGroup::class.java)
                intent.putExtra("org.tiqr.challenge", challenge)
                intent.putExtra("org.tiqr.protocolVersion", "2")
                startActivity(intent)
            }

            override fun onParseEnrollmentChallengeError(error: ParseEnrollmentChallengeError) {
                activityDialog?.cancel()

                AlertDialog.Builder(this@MainActivity)
                        .setTitle(error.title)
                        .setMessage(error.message)
                        .setCancelable(false)
                        .setPositiveButton(R.string.ok_button, null)
                        .show()
            }
        })
    }

    companion object {
        private const val MY_PERMISSIONS_REQUEST_CAMERA = 42
    }
}
